### ISSUE TITLE
Non integer labels in Colin27 2008 version

### ISSUE THREAD

**Issue by GFabien** [fepegar/torchio#220]
**🐛Bug**

Class labels in Colin27 dataset, version 2008, are not integer values.

**To reproduce**

```python
from torchio.datasets import Colin27
colin = Colin27(2008)
label_values = colin['cls']['data'].unique()
print(label_values)

>>> tensor([ 0.0000,  0.9882,  2.0235,  3.0118,  4.0000,  4.9882,  6.0235,  7.0118,
         8.9882, 10.0235, 11.0118, 12.0000])
```

**Expected behavior**

Label encoding map should only contain integer values.

**TorchIO version**

0.17.7

**Comment by fepegar** [comment]
This is weird, and I don't think there's much we can do about this. I tried reading the image with ITK and results are the same. I checked the intercept and slope values in the NIfTI header and I found nothing interesting:

```
scl_slope       : nan
scl_inter       : nan
```

I've been googling with no success. I think that the only solution for now is to `round` the data in the `Colin27` class of the `torchio.datasets.mni.colin` module.

**Comment by fepegar** [comment]
I'll check the MINC versions.

**Comment by fepegar** [comment]
> I'll check the MINC versions.

I got the same results.

**Comment by fepegar** [comment]
Might be related to the **Small float mantissa precision – concrete example** section in https://bartwronski.com/

**Comment by fepegar** [comment]
In the NIfTI header it says `int16`, but the for some reason the data is loaded as `float64` even when I do

```python
np.asanyarray(nii.dataobj)
```

**Comment by romainVala** [comment]
mrinfo say there is an intensity scaling
  Data type:         signed 16 bit integer (little endian)
  Intensity scaling: offset = 0, multiplier = 0.047058824449777603
strange that you do not see the intercept and slope (I thought it was related)

**Comment by fepegar** [comment]
@romainVala, do you know how/if that multiplier is meant to be used?

**Comment by GFabien** [comment]
From nibabel:
```python
image.dataobj.slope
>>> 0.0470588244497776
```
But also from nibabel:
```python
nib.analyze.get_slope_inter(image)
>>> (1.0, 0.0)
```

**Comment by fepegar** [comment]
Note that this is a NIfTI image, not analyze:

```python
nii.header.get_slope_inter()
```

**Comment by fepegar** [comment]
Some more info about this stuff, for reference: https://nipy.org/nibabel/nifti_images.html

**Comment by fepegar** [comment]
> In the NIfTI header it says int16, but for some reason the data is loaded as float64

I suspect this happens because there are values for intercept and slope in the original file.

**Comment by fepegar** [comment]
Maybe those values shouldn't be there in the first place. I'll try modifying the file directly to remove these values and loading it with nibabel.

**Comment by fepegar** [comment]
I removed the slope from the file

```python
In [7]: with open('colin27_cls_tal_hires.nii', 'rb+') as f:
   ...:     f.seek(116)  # address of intercept, which I know is 0
   ...:     data = f.read(4)  # stored in float32 so 4 bytes
   ...:     f.seek(112)  # address of slope
   ...:     f.write(data)  # actually this should have been 1, but it looks like nibabel uses a 1 if it finds a 0
   ...:
```

and the values are now

```python
In [14]: np.unique(lm)
Out[14]:
array([  0,  21,  43,  64,  85, 106, 128, 149, 191, 213, 234, 255],
      dtype=int16)
```

Indeed to get 12 from 255 you need to multiply by `12/255 = 0.047058823529411764`, which is the intercept stored in the header. However, `234 * 0.047058823529411764 = 11.0118`. It looks like, for some reason, someone decided to store the labels from 0 to 255 for some reason, and adding a slope. Maybe this worked with the software they use at the MNI because the labels are rounded after reading.

I think the best for now is just rounding the values ourselves when reading the CLS image.

**Comment by fepegar** [comment]
(I edited this comment, forgot to add the returned values: https://github.com/fepegar/torchio/issues/220#issuecomment-656100610)

**Comment by fepegar** [comment]
Fixed (hacked) in `v0.17.9`. Thanks for reporting and investigating! I'm going to email the person in charge of the atlas to ask about this.
